docs/how_pool_works.md

```diff
@@ -0,0 +1,198 @@
+
+There are 3 roles present in the contract.
+Any Ethereum address can simultaneously take on any of these roles.
+
+- **The sender**: has assets and chooses who do they want to send them to,
+how much, and at what rate
+- **The receiver:** receives funds from senders
+- **The proxy**: receives funds from senders,
+but immediately passes them to receivers of their choice
+
+## The cycles
+
+The whole blockchain history is divided into cycles of equal length.
+Every cycle contains the same, constant number of blocks.
+In the examples below, we assume that the cycle length is 5.
+
+![](how_pool_works_1.png)
+
+# The receiver
+
+There are no setup steps for one to become a receiver.
+Any address can receive donations at any time, from any sender.
+The only function of this role is collection of funds sent by others.
+
+## The deltas
+
+On every block, a sender can send funds to a receiver.
+The receiver can collect funds sent on a given block only when the cycle containing it is over.
+
+![](how_pool_works_2.png)
+
+Here, we see the timeline of a receiver who is receiving funds from two senders.
+Each of the senders has sent different amounts over different periods of time.
+At the end of each cycle, the collectable amount was increased by the total sent amount.
+
+Each receiver needs to know, how much was sent to them.
+To accomplish that the history of received funds needs to be stored in some form.
+The receiver doesn't care, who exactly is sending on each block and in what configuration.
+The only thing that matters are the collectable per-cycle values below the timeline.
+This reduces the amount of stored data significantly, but this can be optimized further.
+
+![](how_pool_works_3.png)
+
+In this example, we started with having the raw collectable value of 23 for every block until
+the end of time.
+We then reduced that to storing values added to the collectable amount on each cycle.
+Now we need to describe only cycles when receiving anything.
+The senders usually are sending constant per-cycle amounts over long periods of time, so
+the added values tend to create long series of constant numbers, in this case, 5s.
+We exploit that and turn them into deltas relative to the previous cycles.
+Now we need deltas only for cycles where the sending rate changes, it's very cheap.
+This is what the contract actually stores: a mapping from cycle numbers to deltas.
+
+## Collecting
+
+When the receiver wants to collect the funds, they can iterate over the finished cycles, reconstruct
+the added values from deltas and then calculate the collectable amount at this point in time.
+For any cycle before the first funding begins the added value is assumed to be 0.
+
+# The sender
+
+The sender has an available balance, a per-block rate, and a set of receivers.
+From the available balance, the per-block rate is disbursed across all the receivers on every block.
+
+The balance can be increased by topping up, which requires sending the assets to the contract.
```

Comment:
We should clearly point out the actions that a sender can take and mention that explicitly instead of using passive voice.

Reply:
Fixed